File: examples/cpp/hand/agg/handlers/post_blind_handler.cpp
#include "post_blind_handler.hpp"
#include "angzarr/errors.hpp"
#include <chrono>
#include <google/protobuf/util/time_util.h>

namespace hand {
namespace handlers {

examples::BlindPosted handle_post_blind(
    const examples::PostBlind& cmd,
    const HandState& state) {

    // Guard
    if (!state.exists()) {
        throw angzarr::CommandRejectedError::not_found("Hand not dealt");
    }
    if (state.status == "complete") {
        throw angzarr::CommandRejectedError::precondition_failed("Hand is complete");
    }

    // Validate
    if (cmd.player_root().empty()) {
        throw angzarr::CommandRejectedError::invalid_argument("player_root is required");
    }

    const PlayerHandInfo* player = state.get_player(cmd.player_root());
    if (!player) {
        throw angzarr::CommandRejectedError::not_found("Player not in hand");
    }
    if (player->has_folded) {
        throw angzarr::CommandRejectedError::precondition_failed("Player has folded");
    }
    if (cmd.amount() <= 0) {
        throw angzarr::CommandRejectedError::invalid_argument("Blind amount must be positive");
    }

    // Compute
    int64_t actual_amount = std::min(cmd.amount(), player->stack);
    int64_t new_stack = player->stack - actual_amount;
    int64_t new_pot_total = state.get_pot_total() + actual_amount;

    auto now = std::chrono::system_clock::now();
    auto timestamp = google::protobuf::util::TimeUtil::TimeTToTimestamp(
        std::chrono::system_clock::to_time_t(now));

    examples::BlindPosted event;
    event.set_player_root(cmd.player_root());
    event.set_blind_type(cmd.blind_type());
    event.set_amount(actual_amount);
    event.set_player_stack(new_stack);
    event.set_pot_total(new_pot_total);
    *event.mutable_posted_at() = timestamp;

    return event;
}

} // namespace handlers
} // namespace hand

File: examples/csharp/Hand/Agg/Handlers/PostBlindHandler.cs
using Google.Protobuf.WellKnownTypes;
using Angzarr.Client;
using Angzarr.Examples;

namespace Hand.Agg.Handlers;

/// <summary>
/// Handler for PostBlind command.
/// </summary>
public static class PostBlindHandler
{
    public static BlindPosted Handle(PostBlind cmd, HandState state)
    {
        // Guard
        if (!state.Exists)
            throw CommandRejectedError.PreconditionFailed("Hand not dealt");
        if (state.Status == "complete")
            throw CommandRejectedError.PreconditionFailed("Hand is complete");

        // Validate
        if (cmd.PlayerRoot.IsEmpty)
            throw CommandRejectedError.InvalidArgument("player_root is required");

        var player = state.GetPlayer(cmd.PlayerRoot);
        if (player == null)
            throw CommandRejectedError.PreconditionFailed("Player not in hand");
        if (player.HasFolded)
            throw CommandRejectedError.PreconditionFailed("Player has folded");
        if (cmd.Amount <= 0)
            throw CommandRejectedError.InvalidArgument("Blind amount must be positive");

        // Compute
        var actualAmount = Math.Min(cmd.Amount, player.Stack);
        var newStack = player.Stack - actualAmount;
        var newPotTotal = state.GetPotTotal() + actualAmount;

        return new BlindPosted
        {
            PlayerRoot = cmd.PlayerRoot,
            BlindType = cmd.BlindType,
            Amount = actualAmount,
            PlayerStack = newStack,
            PotTotal = newPotTotal,
            PostedAt = Timestamp.FromDateTime(DateTime.UtcNow)
        };
    }
}
